codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -137,145 +207,104 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
+            childChannel.streamClosed();
         }
     }
 
-    private void onStreamActive(int streamId, Http2HeadersFrame headersFrame) {
-        final Http2StreamChannel childChannel;
-        if (isOutboundStream(server, streamId)) {
-            if (!(headersFrame instanceof ChannelCarryingHeadersFrame)) {
-                throw new IllegalArgumentException("needs to be wrapped");
-            }
-            childChannel = ((ChannelCarryingHeadersFrame) headersFrame).channel();
-            childChannel.streamId(streamId);
-        } else {
-            ChannelFuture future = bootstrap.connect(streamId);
-            childChannel = (Http2StreamChannel) future.channel();
-        }
+    private DefaultHttp2StreamChannel onStreamActive(Http2FrameStream stream) {
+        DefaultHttp2StreamChannel childChannel = channels.get(stream);
+        if (childChannel == null) {
+            childChannel = newStreamChannel(stream);
 
-        Http2StreamChannel existing = childChannels.put(streamId, childChannel);
-        assert existing == null;
-    }
+            childChannel.pipeline().addLast(inboundStreamHandler);
 
-    private void onStreamClosed(int streamId) {
-        final Http2StreamChannel childChannel = childChannels.remove(streamId);
-        if (childChannel != null) {
-            final EventLoop eventLoop = childChannel.eventLoop();
-            if (eventLoop.inEventLoop()) {
-                onStreamClosed0(childChannel);
-            } else {
-                eventLoop.execute(new Runnable() {
-                    @Override
-                    public void run() {
-                        onStreamClosed0(childChannel);
-                    }
-                });
-            }
+            ChannelFuture future = ctx.channel().eventLoop().register(childChannel);
+            future.addListener(CHILD_CHANNEL_REGISTRATION_LISTENER);
         }
-    }
-
-    private void onStreamClosed0(Http2StreamChannel childChannel) {
-        assert childChannel.eventLoop().inEventLoop();
 
-        childChannel.onStreamClosedFired = true;
-        childChannel.fireChildRead(CLOSE_MESSAGE);
+        assert !childChannel.isWritable();
+        childChannel.incrementOutboundFlowControlWindow(initialOutboundStreamWindow);
+        childChannel.pipeline().fireChannelWritabilityChanged();
+        return childChannel;
     }
 
-    void flushFromStreamChannel() {
-        EventExecutor executor = ctx.executor();
-        if (executor.inEventLoop()) {
-            flush(ctx);
-        } else {
-            Runnable task = flushTask;
-            if (task == null) {
-                task = flushTask = new Runnable() {
-                    @Override
-                    public void run() {
-                        flush(ctx);
-                    }
-                };
-            }
-            executor.execute(task);
-        }
+    // TODO: This is most likely not the best way to expose this, need to think more about it.
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)